[docs] Improve sorting documentation page (#3564)
mui-org/material-ui-x: docs/src/pages/components/data-grid/sorting/sorting.md

# Data Grid - Sorting

<p class="description">Easily sort your rows based on one or several criteria.</p>

Sorting is enabled by default to the grid users and works out of the box without any explicit configuration.
Users can set a sorting rule simply by clicking on a column header.
Following clicks change the column's sorting direction. You can see the applied direction on the header's arrow indicator.

{{"demo": "pages/components/data-grid/sorting/BasicExampleDataGrid.js", "bg": "inline", "defaultCodeOpen": false}}

## Single and multi-sorting

> > ⚠️ The `DataGrid` can only sort the rows according to one criterion at the time.

Reviewer: ```suggestion
> ⚠️ The `DataGrid` can only sort the rows according to one criterion at a time.
```

# Data Grid - Sorting

<p class="description">Easily sort your rows based on one or several criteria.</p>

> The sorting rules can be modified through the grid interface by clicking on a column header. Repeat this action to pass to the next sorting direction of the column.

Reviewer 2: To make easier to add suggestions.

```suggestion
The sorting rules can be modified through the user interface by clicking on a column header.
Repeat this action to pass to the next sorting direction of the column.
```

{{"demo": "pages/components/data-grid/sorting/BasicExampleDataGrid.js", "bg": "inline", "defaultCodeOpen": false}}

> ## Single and multi-sorting

Reviewer 3: This section content is only about multi-sorting
```suggestion
## Single and multi-sorting [<span class="plan-pro"></span>](https://mui.com/store/items/material-ui-pro/)
```

Reviewer: Thanks. I think the preview is looking great.

> ⚠️ The `DataGrid` can only sort the rows according to one criterion at the time.
>
> To use multi-sorting, you need to upgrade to the [Pro plan](https://mui.com/store/items/material-ui-pro/)

> The demo below let you sort the rows according to several criteria at the same time.

Reviewer 3: ```suggestion
The demo below lets you sort the rows according to several criteria at the same time.
```

Hold down the <kbd class="key">CTRL</kbd> or <kbd class="key">Shift</kbd> (use <kbd class="key">⌘ Command</kbd> on macOS) key while clicking the column header.

{{"demo": "pages/components/data-grid/sorting/BasicExampleDataGridPro.js", "bg": "inline", "defaultCodeOpen": false}}

## Pass sorting rules to the grid

### Structure of the model

> The sort model is a list of sorting item.

Reviewer 3: ```suggestion
The sort model is a list of sorting items.
```

> Each item represent a sorting rule and is composed of several elements:

Reviewer 3: ```suggestion
Each item represents a sorting rule and is composed of several elements:
```

- `sortingItem.field`: the field on which we want to apply the rule

> - `sortingItem.sort`: the direction of the sorting (`'asc'`, `'desc'`, `null` or `undefined`, if `null` or `undefined`, the rule will not be applied)

Reviewer 3: ```suggestion
- `sortingItem.sort`: the direction of the sorting (`'asc'`, `'desc'`, `null` or `undefined`). If `null` or `undefined`, the rule will not be applied.
```
I'm wondering if "`null` or `undefined`" should be in the parenthesis

Author: I alway see them with ````

> 

Reviewer 3: After reading it, I'm wondering what append if I set an array with more than one item in the MIT version

Author: We should unify with the filtering
For the filtering, we prevent from sorting several items in the state
For the sorting, we prevent when updating from the UI, but not from the `initialState` or `sortModel` which is a major issue...

See: https://codesandbox.io/s/datagrid-v5-quick-start-forked-olg5j?file=/src/App.tsx

More globally, I think the `DataGrid` checks should be done in the `setXXXModel` method whenever possible to avoid problems.
The prop types checks can be interesting sometime, but it's too easy to fool.

Author: See #3671

### Initialize the sort model

To initialize the sort model without controlling it, provide the model to the `initialState` prop.

```jsx
<DataGrid
  initialState={{
    sorting: {
      sortModel: [{ field: 'rating', sort: 'desc' }],
    },
  }}
/>
```

{{"demo": "pages/components/data-grid/sorting/InitialSort.js", "bg": "inline", "defaultCodeOpen": false}}

### Controlled sort model

Use the `sortModel` prop to control the sorting rules applied on the rows.

> You can use the `onSortModelChange` prop to listen to changes to the sorting rules and update the prop accordingly.

Reviewer 3: ```suggestion
You can use the `onSortModelChange` prop to listen to changes in the sorting rules and update the prop accordingly.
```
To avoid to to to

{{"demo": "pages/components/data-grid/sorting/ControlledSort.js", "bg": "inline", "defaultCodeOpen": false}}

> ## Disable the sorting

Reviewer 3: I think this section could be introduced before multi sorting because: It does not need to know the existence of multi-sorting, and it is simpler to understand than sorting model

Author: Would you have the same logic for filtering ? 
I kept the same order between the two pages as the logic are similar.

Reviewer 3: I don't know.

Having the same structure is clearly a good point (I was wondering why the flow seems familiar 😅)

I noticed it for this page (and not the filtering one) because here goes from a section with 3 subsections to a small section with one sentence + an example.

Let see if @joserodolfofreitas and @cherniavskii have an opinion

Author: Not sure to follow you here

> or remove the configuration altogether

What configuration are we talking about ?

Reviewer: You're right, there's no configuration on the `single sorting` and `multi sorting` basic demos (great!). I may have mixed it with another one that sets a default sorting rule (which was the config I was referring to).

Author: You can disable the multi sorting on the pro
But I think this option mainly exist to limit on the free, we don't even document it :+1:

By default, all columns are sortable.
To disable the filter on a column, set the `sortable` property of `GridColDef` to `false`.
In the example below, the _rating_ column can not be sorted.

```tsx
<Datagrid columns={[...columns, { field: 'rating', sortable: false }]} />
```

> {{"demo": "pages/components/data-grid/sorting/DisableSortingGrid.js", "bg": "inline"}}

Reviewer 3: ```suggestion
{{"demo": "pages/components/data-grid/sorting/DisableSortingGrid.js", "bg": "inline", "defaultCodeOpen": false}}
```

## Custom comparator

A comparator determines how two cell values should be sorted.

Each column type comes with a default comparator method.
You can re-use them by importing the following functions:

- `gridStringNumberComparator` (used by the `string` and `singleSelect` columns)
- `gridNumberComparator` (used by the `number` and `boolean` columns)
- `gridDateComparator` (used by the `date` and `date-time` columns)

To extend or modify this behavior in a specific column, you can pass in a custom comparator, and override the `sortComparator` property of the `GridColDef` interface.

In the example below:

> - the "Name" column combines the `name` and `isAdmin` fields. The sorting is based on `isAdmin` and then on `name` if necessary. It re-uses built-in sorting comparator.

Reviewer 2: ```suggestion
- the "Name" column combines the `name` and `isAdmin` fields. The sorting is based on `isAdmin` and then on `name`, if necessary. It re-uses the built-in sorting comparator.
```

- the "Created on" column sorting is based on the day of the month of the `createdOn` field. It is a fully custom sorting comparator.

> {{"demo": "pages/components/data-grid/sorting/ComparatorSortingGrid.js", "bg": "inline"}}

Reviewer 3: ```suggestion
{{"demo": "pages/components/data-grid/sorting/ComparatorSortingGrid.js", "bg": "inline", "defaultCodeOpen": false}}
```

The preview does not bring any useful information

The sort model is a list of sorting items.
Each item represents a sorting rule and is composed of several elements:

- `sortingItem.field`: the field on which we want to apply the rule
- `sortingItem.sort`: the direction of the sorting (`'asc'`, `'desc'`, `null` or `undefined`). If `null` or `undefined`, the rule will not be applied.

### Initialize the sort model

To initialize the sort model without controlling it, provide the model to the `initialState` prop.

```jsx
<DataGrid
  initialState={{
    sorting: {
      sortModel: [{ field: 'rating', sort: 'desc' }],
    },
  }}
/>
```

{{"demo": "pages/components/data-grid/sorting/InitialSort.js", "bg": "inline", "defaultCodeOpen": false}}

### Controlled sort model

Use the `sortModel` prop to control the sorting rules applied on the rows.

You can use the `onSortModelChange` prop to listen to changes in the sorting rules and update the prop accordingly.

{{"demo": "pages/components/data-grid/sorting/ControlledSort.js", "bg": "inline", "defaultCodeOpen": false}}

## Disable the sorting

By default, all columns are sortable.
To disable the filter on a column, set the `sortable` property of `GridColDef` to `false`.

> In the example below, the _rating_ column can not be sorted.

Reviewer: ```suggestion
In the example below, the user cannot use the _rating_ column as a sorting rule.
```

```tsx
<Datagrid columns={[...columns, { field: 'rating', sortable: false }]} />
```

{{"demo": "pages/components/data-grid/sorting/DisableSortingGrid.js", "bg": "inline", "defaultCodeOpen": false}}

## Custom comparator

A comparator determines how two cell values should be sorted.

Each column type comes with a default comparator method.
You can re-use them by importing the following functions:

- `gridStringNumberComparator` (used by the `string` and `singleSelect` columns)
- `gridNumberComparator` (used by the `number` and `boolean` columns)
- `gridDateComparator` (used by the `date` and `date-time` columns)

To extend or modify this behavior in a specific column, you can pass in a custom comparator, and override the `sortComparator` property of the `GridColDef` interface.

In the example below:

- the "Name" column combines the `name` and `isAdmin` fields. The sorting is based on `isAdmin` and then on `name`, if necessary. It re-uses the built-in sorting comparator.

> - the "Created on" column sorting is based on the day of the month of the `createdOn` field. It is a fully custom sorting comparator.

Reviewer: I feel we should break this into **two** examples.
Perhaps the first showing only the fully custom comparator, and then a second showing how to extend it using the built-in methods.

In practice, when you click a column that is not sorted, it will sort ascending (`asc`).
The next click will make it sort descending (`desc`). Another click will remove the sort (`null`), reverting to the order that the data was provided in.

### For all columns

> The default sort order can be overwritten for all columns with the `sortingOrder` prop.

Reviewer 2: ```suggestion
The default sort order can be overridden for all columns with the `sortingOrder` prop.
```

];
```

> {{"demo": "pages/components/data-grid/sorting/OrderSortingPerColumnGrid.js", "bg": "inline", "defaultCodeOpen": false}}

Reviewer 3: The example on top of it should be updated "quantity" column is not visible anymore

Reviewer 3: Same remark in the code of the demonstration

## Server-side sorting

> Sorting can be run service-side by setting the `sortingMode` prop to `server`, and implementing the `onSortModelChange` handler.

Reviewer 2: ```suggestion
Sorting can be run server-side by setting the `sortingMode` prop to `server`, and implementing the `onSortModelChange` handler.
```
